rust/lance/src/index/vector/ivf/v2.rs

```diff
@@ -0,0 +1,508 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+//! IVF - Inverted File index.
+
+use core::fmt;
+use std::{
+    any::Any,
+    collections::HashMap,
+    sync::{Arc, Weak},
+};
+
+use arrow::{
+    array::as_struct_array,
+    compute::{concat_batches, sort_to_indices, take},
+};
+use arrow_arith::numeric::sub;
+use arrow_array::{RecordBatch, StructArray, UInt32Array};
+use async_trait::async_trait;
+use deepsize::DeepSizeOf;
+use futures::prelude::stream::{self, StreamExt, TryStreamExt};
+use lance_core::{cache::DEFAULT_INDEX_CACHE_SIZE, Error, Result};
+use lance_file::v2::reader::FileReader;
+use lance_index::{
+    pb,
+    vector::{
+        ivf::storage::IVF_METADATA_KEY,
+        quantizer::Quantization,
+        v3::{storage::IvfQuantizationStorage, subindex::IvfSubIndex, DISTANCE_TYPE_KEY},
+        Query, DIST_COL,
+    },
+    Index, IndexType, INDEX_AUXILIARY_FILE_NAME, INDEX_FILE_NAME,
+};
+use lance_io::{
+    object_store::ObjectStore, scheduler::ScanScheduler, traits::Reader, ReadBatchParams,
+};
+use lance_linalg::{distance::DistanceType, kernels::normalize_arrow};
+use moka::sync::Cache;
+use object_store::path::Path;
+use prost::Message;
+use roaring::RoaringBitmap;
+use snafu::{location, Location};
+use tracing::instrument;
+
+use crate::{
+    index::{vector::VectorIndex, PreFilter},
+    session::Session,
+};
+
+use super::{centroids_to_vectors, Ivf, IvfIndexPartitionStatistics, IvfIndexStatistics};
+/// IVF Index.
+#[derive(Debug)]
+pub struct IVFIndex<I: IvfSubIndex + 'static, Q: Quantization> {
+    uuid: String,
+
+    /// Ivf model
+    ivf: Ivf,
+
+    reader: FileReader,
+    storage: IvfQuantizationStorage<Q>,
+
+    /// Index in each partition.
+    sub_index_cache: Cache<String, Arc<I>>,
+
+    distance_type: DistanceType,
+
+    // The session cache holds an Arc to this object so we need to
+    // hold a weak pointer to avoid cycles
+    /// The session cache, used when fetching pages
+    #[allow(dead_code)]
+    session: Weak<Session>,
+}
+
+impl<I: IvfSubIndex, Q: Quantization> DeepSizeOf for IVFIndex<I, Q> {
+    fn deep_size_of_children(&self, context: &mut deepsize::Context) -> usize {
+        self.uuid.deep_size_of_children(context)
+        // Skipping session since it is a weak ref
+    }
```

Comment:
What about the other properties?  E.g. `storage`, `sub_index_cache`, etc.?

Reply:
added storage, `sub_index_cache` is now for passing the the object safe requirement (we can't cache the partition in session's cache now), it will be removed soon